### ISSUE TITLE
"extract_all_from_idf_file" might throw error if IDF contains mutiple week rules

### ISSUE THREAD

**Issue by zha** [ladybug-tools/honeybee-energy#56]
I was trying to execute `extract_all_from_idf_file` for the following IDF command but failed. It gives me the error "ScheduleDay objects can be assigned to a ScheduleRuleset only once."
```
ScheduleTypeLimits,
  avail_limit,                 !- Name
  0,                                      !- Lower Limit Value {BasedOnField A3}
  1,                                      !- Upper Limit Value {BasedOnField A3}
  Discrete,                               !- Numeric Type
  availability;                           !- Unit Type


Schedule:Day:Interval,
  off,                                   !- Name
  avail_limit,                 !- Schedule Type Limits Name
  No,                                     !- Interpolate to Timestep
  24:00,                                  !- Time 1 {hh:mm}
  0;                                      !- Value Until Time 1
Schedule:Day:Interval,
  on,                                   !- Name
  avail_limit,                 !- Schedule Type Limits Name
  No,                                     !- Interpolate to Timestep
  24:00,                                  !- Time 1 {hh:mm}
  1;                                      !- Value Until Time 1


Schedule:Week:Daily, 
Weeks_always_off,                       !- Name
off,                                    !- Sunday Schedule:Day Name
off,                                    !- Monday Schedule:Day Name
off,                                    !- Tuesday Schedule:Day Name
off,                                    !- Wednesday Schedule:Day Name
off,                                    !- Thursday Schedule:Day Name
off,                                    !- Friday Schedule:Day Name
off,                                    !- Saturday Schedule:Day Name
off,                                    !- Holiday Schedule:Day Name
off,                                    !- SummerDesignDay Schedule:Day Name
off,                                    !- WinterDesignDay Schedule:Day Name
off,                                    !- CustomDay1 Schedule:Day Name
off;                                    !- CustomDay2 Schedule:Day Name


Schedule:Week:Daily,
Week_cooling_switch_on,                        !- Name
off,                                     !- Sunday Schedule:Day Name
off,                                     !- Monday Schedule:Day Name
on,                                    !- Tuesday Schedule:Day Name
on,                                    !- Wednesday Schedule:Day Name
on,                                    !- Thursday Schedule:Day Name
on,                                    !- Friday Schedule:Day Name
on,                                    !- Saturday Schedule:Day Name
off,                                    !- Holiday Schedule:Day Name
off,                                    !- SummerDesignDay Schedule:Day Name
off,                                    !- WinterDesignDay Schedule:Day Name
off,                                    !- CustomDay1 Schedule:Day Name
off;                                    !- CustomDay2 Schedule:Day Name


Schedule:Year,
cooling_avail,                          !- Name
avail_limit,                 !- Schedule Type Limits Name
Weeks_always_off,   !- Schedule:Week Name 1
1,                                      !- Start Month 1
1,                                      !- Start Day 1
4,                                      !- End Month 1
2,                                      !- End Day 1

Week_cooling_switch_on,   !- Schedule:Week Name 1
4,                                      !- Start Month 1
3,                                      !- Start Day 1
12,                                      !- End Month 1
31;                                     !- End Day 1

```

**Comment by chriswmackey** [comment]
Thank you for reporting this @zha and for the clear example that recreates the issue (I've been able to recreate it on my end). 
This looks like a bug that results from the fact that we changed the rules for ScheduleDays to avoid people overwriting ScheduleDay properties when they are referenced by multiple WeekSchedules.
A simple duplication of the cross-referenced ScheduleDay object should fix this and prevent such overwriting. I will try to fix it now.

**Comment by chriswmackey** [comment]
I have put together a fix in that PR there, @zha . I would be curious in know what use case you are using the `extract_all_from_idf_file` method for and whether the solution we arrived at will still work for that case.

**Comment by zha** [comment]
Dear @chriswmackey and @mostaphaRoudsari 

Sorry for the delayed response and thank you for taking care of this issue. It seems that there is no error now.

I did not take a closer look at the codes but it seems that you approached this error by attaching some random string sequence to the `DaySchedule` name instead of corss-referencing each other? I see where you are coming from and I am OK with it, but my concern is that it MIGHT make the IDF file difficult to read since it can make the IDF string pretty long by the end of the day. Just my $0.02

I spotted this error when I was trying to define the heating and cooling availability schedules for the ideal air object. I am in Toronto so the HTG/CLG schedules in my simulation should follow Toronto's rental by-law. The HTG/CLG schedules have four unique week rules: weeks always-on; weeks always off; week switching on; and week switching off, and all of them reference the same day rules: On and Off. I already had the raw IDF files handy (https://github.com/zha/early-building-design-assessment/blob/master/design_simulation/dat/CLG_availiability.idf), so I attempted to read in the files using the `extract_all_from_idf_file` method and that's when I found the error. But I moved on with my own mod and everything appears working fine now.

Also worth mentioning: it looks like the `IdealAirSystem` does not currently support defining heating and cooling availability schedule? Anyway, I implemented it on my own and here they are: 28d9a54d7c40e2637a3f9099823c17e86120f1df 95b570ae35e99c0dda4c255c419febbd246f9934 (apologize in advance if my codes are hard to read. I am not a seasoned programmer after all)

Also reference 
 the pull request: https://github.com/ladybug-tools/honeybee-energy/pull/57 

Thank you very much again! ^_^

Shen

**Comment by chriswmackey** [comment]
Thanks for your input @zha .  I ended up [implementing a solution that better supports the workflow that you were using](https://github.com/ladybug-tools/honeybee-energy/pull/57) (so the names of the ScheduleDays won't get changed) and you can have the same ScheduleDay instance used multiple times within a single ScheduleRulest/ScheduleYear.  If you can't tell, this honeybee-energy library is still undergoing some refactoring and it isn't used by any of the Grasshopper plugins yet.  While this means that there may be bugs like this one and missing features (like availability schedules), your input now helps us make big structural changes that we probably won't be able to make once this library is used by a Grasshopper plugin.  So we are very happy to  have your thoughts.

BTW, I opened [a separate issue for the availability schedules](https://github.com/ladybug-tools/honeybee-energy/issues/58) and I'll get around to this after I have the pathway from honeybee-energy to openstudio working well.
